### ISSUE TITLE
Long startup time due to metabase.sync running on startup for every database

### ISSUE THREAD

**Issue by bchrobot** [metabase/metabase#19604]
**Describe the bug**

We have a Metabase database per-client (currently 550). Metabase runs the full `metabase.sync` set of functions for each database in series on startup. This leads to very long (8+ hour) startup times.

**Logs**

Nothing out of the ordinary, just the expected `metabase.sync`

<details>
<summary>Logs</summary>

```
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.task.sync-databases Starting sync task for Database 772.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util STARTING: Sync metadata for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util STARTING: step 'sync-timezone' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util FINISHED: step 'sync-timezone' for postgres Database 772 '********' (4.0 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util STARTING: step 'sync-tables' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util FINISHED: step 'sync-tables' for postgres Database 772 '********' (65.7 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 INFO metabase.sync.util STARTING: step 'sync-fields' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'name' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'name' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'oid' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'oid' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'oid' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'name' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:16-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type 'name' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_int4' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_text' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_text' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_text' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_int4' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_text' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 WARN metabase.driver.sql-jdbc.sync.describe-table Don't know how to map column type '_int4' to a Field base_type, falling back to :type/*.
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 INFO metabase.sync.util FINISHED: step 'sync-fields' for postgres Database 772 '********' (1.4 s)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:17-05:00 INFO metabase.sync.util STARTING: step 'sync-fks' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util FINISHED: step 'sync-fks' for postgres Database 772 '********' (751.7 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util STARTING: step 'sync-metabase-metadata' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util FINISHED: step 'sync-metabase-metadata' for postgres Database 772 '********' (79.5 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util FINISHED: Sync metadata for postgres Database 772 '********' (2.3 s)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util STARTING: Analyze data for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.util STARTING: step 'fingerprint-fields' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.analyze fingerprint-fields Analyzed [*****·············································] 😢   10% Table 11,605 'public.assignable_campaign_contacts'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:18-05:00 INFO metabase.sync.analyze fingerprint-fields Analyzed [**********········································] 😞   20% Table 11,582 'public.team'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze fingerprint-fields Analyzed [***************···································] 😕   31% Table 17,932 'public.external_sync_config_question_response_result_code'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util FINISHED: step 'fingerprint-fields' for postgres Database 772 '********' (534.8 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util STARTING: step 'classify-fields' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-fields Analyzed [********************······························] 😐   41% Table 11,600 'public.messaging_service_stick'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-fields Analyzed [*************************·························] 😬   51% Table 11,577 'public.campaign'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-fields Analyzed [******************************····················] 😌   61% Table 17,927 'public.missing_external_sync_question_response_configuration'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util FINISHED: step 'classify-fields' for postgres Database 772 '********' (44.1 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util STARTING: step 'classify-tables' for postgres Database 772 '********'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-tables Analyzed [***********************************···············] 😋   72% Table 11,594 'public.assignable_needs_reply'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-tables Analyzed [****************************************··········] 😊   82% Table 11,616 'public.link_domain'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.analyze classify-tables Analyzed [**********************************************····] 😍   92% Table 11,573 'public.assignment'
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util FINISHED: step 'classify-tables' for postgres Database 772 '********' (52.6 ms)
[c5b20228-ab6e-4e19-a9cf-8201bfab714b] 2022-01-10T08:23:19-05:00 INFO metabase.sync.util FINISHED: Analyze data for postgres Database 772 '********' (643.7 ms)
```

</details>

**To Reproduce**
1. Add many databases to Metabase
2. Restart Metabase
3. See error

**Expected behavior**

A configuration option, envvar or application, for deferring `metabase.sync` or making it a non-blocking part of startup.

**Screenshots**

N/A - just the loading screen.

**Information about your Metabase Installation:**

<details>

<summary>Diagnostic Info</summary>

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.13+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.13",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.13+8",
    "os.name": "Linux",
    "os.version": "5.4.120+",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "clickhouse"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "13.4 (Ubuntu 13.4-4.pgdg18.04+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.23"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-12-16",
      "tag": "v0.41.5",
      "branch": "release-x.41.x",
      "hash": "fbfffc6"
    },
    "settings": {
      "report-timezone": "America/New_York"
    }
  }
}
```

</details>

- Your browser and the version: Brave Version 1.33.106 Chromium: 96.0.4664.110 (Official Build) (x86_64)
- Your operating system: macOS 12.1
- Your databases: Postgres, ClickHouse
- Metabase version: v0.41.5
- Metabase hosting environment: GKE
- Metabase internal database: Postgres

**Severity**

Mostly annoying, although an HA failover of the Metabase internal database cluster recently required a Metabase restart to reconnect to the master node and client dashboards were down during this time.

**Additional context**

I have searched through open and closed issues, Metabase documentation, and a little of the codebase to try to find my own answer but did not find anything.

**Comment by flamber** [comment]
Please use the forum for questions and troubleshooting: https://discourse.metabase.com/

**Comment by bchrobot** [comment]
> Please use the forum for questions and troubleshooting: https://discourse.metabase.com/

@flamber I accidentally hit "create" before finishing the draft but just finished editing the issue

**Comment by flamber** [comment]
Hi @bchrobot 

Your logs don't show the actual problem or any errors.
When Metabase starts, then it schedules the tasks (sync, scan, etc), but should continue without blocking the startup.

And the tasks will be executed if they are older than what you have configured.
You can disable scan schedule via the GUI: https://www.metabase.com/docs/latest/administration-guide/01-managing-databases.html#choose-when-metabase-syncs-and-scans
To disable sync schedule requires a bit more: https://github.com/metabase/metabase/issues/10398#issuecomment-514412107

I've not seen this problem before, so I'm unsure what is going on or what you are seeing as "error" - please give more details.

**Comment by bchrobot** [comment]
@flamber I haven't seen any error messages in the logs but included them for completeness.

The "error" is that these background jobs appear to be running synchronously as part of startup, leading to very long startup times and blocking use of the application until complete, and asynchronously/non-blocking after startup.

**Comment by flamber** [comment]
@bchrobot Hmm.... seems like it might be triggering slightly earlier than what was meant.
```
2022-01-10 16:33:14,660 INFO db.data-migrations :: Finished running data migrations.
Database setup took 4.0 s
2022-01-10 16:33:14,913 INFO metabase.events :: Starting events listener: :metabase.events.dependencies/Dependencies 👂
2022-01-10 16:33:14,939 INFO metabase.events :: Starting events listener: :metabase.events.activity-feed/ActivityFeed 👂
2022-01-10 16:33:14,945 INFO metabase.events :: Starting events listener: :metabase.events.metabot-lifecycle/MetaBotLifecycle 👂
2022-01-10 16:33:14,949 INFO metabase.events :: Starting events listener: :metabase.events.last-login/LastLogin 👂
2022-01-10 16:33:14,952 INFO metabase.events :: Starting events listener: :metabase.events.notifications/Notifications 👂
2022-01-10 16:33:14,954 INFO metabase.events :: Starting events listener: :metabase.events.sync-database/Sync 👂
2022-01-10 16:33:14,955 INFO metabase.events :: Starting events listener: :metabase.events.driver-notifications/DriverNotifications 👂
2022-01-10 16:33:14,956 INFO metabase.events :: Starting events listener: :metabase.events.revision/Revisions 👂
2022-01-10 16:33:14,957 INFO metabase.events :: Starting events listener: :metabase.events.view-log/ViewLog 👂
2022-01-10 16:33:15,265 INFO metabase.task :: Initializing task SyncDatabases 📆
2022-01-10 16:33:15,410 INFO task.sync-databases :: Starting sync task for Database 8.
2022-01-10 16:33:15,450 INFO sync.util :: STARTING: Sync metadata for mysql Database 8 'mariadb-10.x'
2022-01-10 16:33:15,451 INFO driver.impl :: Initializing driver :sql...
2022-01-10 16:33:15,451 INFO driver.impl :: Initializing driver :sql-jdbc...
2022-01-10 16:33:15,452 INFO driver.impl :: Initializing driver :mysql...
2022-01-10 16:33:15,457 INFO sync.util :: STARTING: step 'sync-timezone' for mysql Database 8 'mariadb-10.x'
2022-01-10 16:33:15,520 INFO task.sync-databases :: Update Field values task triggered for Database 8.
2022-01-10 16:33:15,535 INFO sync.util :: STARTING: Cache field values in mysql Database 8 'mariadb-10.x'
2022-01-10 16:33:15,537 INFO sync.util :: STARTING: step 'update-field-values' for mysql Database 8 'mariadb-10.x'
2022-01-10 16:33:16,477 INFO metabase.task :: Initializing task CheckForNewVersions 📆
2022-01-10 16:33:16,553 INFO metabase.task :: Initializing task SendAnonymousUsageStats 📆
2022-01-10 16:33:16,564 INFO metabase.task :: Initializing task SendAbandomentEmails 📆
2022-01-10 16:33:16,592 INFO metabase.task :: Initializing task SendPulses 📆
2022-01-10 16:33:16,603 INFO metabase.task :: Initializing task SendFollowUpEmails 📆
2022-01-10 16:33:16,621 INFO metabase.task :: Initializing task TaskHistoryCleanup 📆
2022-01-10 16:33:16,639 INFO driver.impl :: Initializing driver :h2...
2022-01-10 16:33:16,651 INFO metabase.core :: Metabase Initialization COMPLETE
```

So a workaround would be to tell Metabase that the next schedule should be in the future, right before starting Metabase.
Something like this will place all sync/scan triggers 60 seconds into the future:
```
UPDATE qrtz_triggers
SET prev_fire_time=(60+extract(epoch from now()))*1000, next_fire_time=(60+extract(epoch from now()))*1000
WHERE trigger_name LIKE 'metabase.task.sync-and-analyze.trigger%' OR trigger_name LIKE 'metabase.task.update-field-values.trigger%';
```

**Comment by bchrobot** [comment]
@flamber ah, that makes sense! Thank you for looking into it and for the query

Tomorrow morning I am going to try adding a kubernetes `initContainer` to the Metabase pod that runs the query. I will report back on whether the init container solution works for kubernetes environment

**Comment by bchrobot** [comment]
@flamber I tried your query with a 15 minute delay instead of 60 seconds. It does successfully delay execution of the sync tasks but does not address the startup length issue.

Given that initialization does eventually complete and execution of SyncDatabase component tasks is happening asynchronously, it seems like [the iteration](https://github.com/metabase/metabase/blob/085496d74602fb7a8e35d11e11f6f83abccf9382/src/metabase/task/sync_databases.clj#L279-L283) over [`check-and-schedule-tasks-for-db`](https://github.com/metabase/metabase/blob/085496d74602fb7a8e35d11e11f6f83abccf9382/src/metabase/task/sync_databases.clj#L202-L236) is what is slow. This is my first foray into Clojure and I don't know it well enough to take it the next step and step through `check-and-schedule-tasks-for-db` to determine what is slow, however.

<details>
<summary>Initial startup logs</summary>

```
Warning: environ value jdk-11.0.13+8 for key :java-version has been overwritten with 11.0.13
WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
2022-01-11 12:40:12,857 INFO metabase.util :: Maximum memory available to JVM: 28.2 GB
2022-01-11 12:40:27,824 DEBUG driver.impl :: Loading driver :sql-jdbc (require metabase.driver.sql-jdbc)
2022-01-11 12:40:27,827 DEBUG driver.impl :: Loading driver :sql (require metabase.driver.sql)
2022-01-11 12:40:27,939 DEBUG native-query-snippet.permissions :: NativeQueryPermissions impl set to metabase.models.native-query-snippet.permissions/default-impl
2022-01-11 12:40:28,987 INFO driver.impl :: Registered abstract driver :sql  🚚
 ↙ Load driver :sql took 1.2 s
2022-01-11 12:40:28,999 INFO driver.impl :: Registered abstract driver :sql-jdbc (parents: [:sql]) 🚚
Load driver :sql-jdbc took 1.2 s
2022-01-11 12:40:29,003 INFO driver.impl :: Registered driver :h2 (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:29,023 INFO driver.impl :: Registered driver :mysql (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:29,744 INFO driver.impl :: Registered driver :postgres (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:31,653 DEBUG integrations.ldap :: LDAP integration set to metabase.integrations.ldap.default_implementation$reify__67186@5baf9daf
2022-01-11 12:40:31,794 DEBUG search.scoring :: Scoring implementation set to metabase.search.scoring$reify__77610@4f1093bd
2022-01-11 12:40:32,174 INFO metabase.core ::
Metabase v0.41.5 (fbfffc6 release-x.41.x)

Copyright © 2022 Metabase, Inc.

Metabase Enterprise Edition extensions are NOT PRESENT.
2022-01-11 12:40:32,184 WARN metabase.core :: WARNING: You have enabled namespace tracing, which could log sensitive information like db passwords.
2022-01-11 12:40:32,185 INFO metabase.core :: Starting Metabase in STANDALONE mode
2022-01-11 12:40:32,227 INFO metabase.server :: Launching Embedded Jetty Webserver with config:
 {:port 3000, :host "0.0.0.0"}

2022-01-11 12:40:32,265 INFO metabase.core :: Starting Metabase version v0.41.5 (fbfffc6 release-x.41.x) ...
2022-01-11 12:40:32,292 INFO metabase.core :: System info:
 {"file.encoding" "UTF-8",
 "java.runtime.name" "OpenJDK Runtime Environment",
 "java.runtime.version" "11.0.13+8",
 "java.vendor" "Eclipse Adoptium",
 "java.vendor.url" "https://adoptium.net/",
 "java.version" "11.0.13",
 "java.vm.name" "OpenJDK 64-Bit Server VM",
 "java.vm.version" "11.0.13+8",
 "os.name" "Linux",
 "os.version" "5.4.120+",
 "user.language" "en",
 "user.timezone" "GMT"}

2022-01-11 12:40:32,293 INFO metabase.plugins :: Loading plugins in /app/plugins...
2022-01-11 12:40:32,937 INFO plugins.dependencies :: Plugin 'Metabase Presto Driver' depends on plugin 'Presto Common Driver'
2022-01-11 12:40:32,939 INFO plugins.dependencies :: Metabase Presto Driver dependency {:plugin Presto Common Driver} satisfied? false
2022-01-11 12:40:32,949 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase Presto Driver"]
2022-01-11 12:40:32,953 INFO plugins.dependencies :: Plugin 'Metabase BigQuery Driver' depends on plugin 'Metabase Google Drivers Shared Dependencies'
2022-01-11 12:40:32,954 INFO plugins.dependencies :: Metabase BigQuery Driver dependency {:plugin Metabase Google Drivers Shared Dependencies} satisfied? false
2022-01-11 12:40:32,955 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase BigQuery Driver" "Metabase Presto Driver"]
2022-01-11 12:40:32,962 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :clickhouse...
2022-01-11 12:40:32,963 INFO driver.impl :: Registered driver :clickhouse (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:32,983 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :snowflake...
2022-01-11 12:40:32,984 INFO driver.impl :: Registered driver :snowflake (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:32,989 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :druid...
2022-01-11 12:40:32,989 INFO driver.impl :: Registered driver :druid  🚚
2022-01-11 12:40:32,994 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :sqlserver...
2022-01-11 12:40:32,995 INFO driver.impl :: Registered driver :sqlserver (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:32,999 INFO plugins.dependencies :: Plugin 'Metabase Google Analytics Driver' depends on plugin 'Metabase Google Drivers Shared Dependencies'
2022-01-11 12:40:33,000 INFO plugins.dependencies :: Metabase Google Analytics Driver dependency {:plugin Metabase Google Drivers Shared Dependencies} satisfied? false
2022-01-11 12:40:33,000 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase BigQuery Driver" "Metabase Presto Driver" "Metabase Google Analytics Driver"]
2022-01-11 12:40:33,007 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :bigquery-cloud-sdk...
2022-01-11 12:40:33,007 INFO driver.impl :: Registered driver :bigquery-cloud-sdk (parents: [:sql]) 🚚
2022-01-11 12:40:33,013 INFO plugins.dependencies :: Metabase cannot initialize plugin Metabase Oracle Driver due to required dependencies. Metabase requires the Oracle JDBC driver in order to connect to Oracle databases, but we can't ship it as part of Metabase due to licensing restrictions. See https://metabase.com/docs/latest/administration-guide/databases/oracle.html for more details.

2022-01-11 12:40:33,013 INFO plugins.dependencies :: Metabase Oracle Driver dependency {:class oracle.jdbc.OracleDriver} satisfied? false
2022-01-11 12:40:33,014 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase BigQuery Driver" "Metabase Presto Driver" "Metabase Oracle Driver" "Metabase Google Analytics Driver"]
2022-01-11 12:40:33,020 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :mongo...
2022-01-11 12:40:33,021 INFO driver.impl :: Registered driver :mongo  🚚
2022-01-11 12:40:33,024 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :presto-common...
2022-01-11 12:40:33,025 INFO driver.impl :: Registered abstract driver :presto-common  🚚
2022-01-11 12:40:33,026 INFO plugins.dependencies :: Metabase Presto Driver dependency {:plugin Presto Common Driver} satisfied? true
2022-01-11 12:40:33,027 DEBUG plugins.initialize :: Dependencies satisfied; these plugins will now be loaded: ["Metabase Presto Driver"]
2022-01-11 12:40:33,027 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :presto...
2022-01-11 12:40:33,028 INFO driver.impl :: Registered driver :presto (parents: [:presto-common]) 🚚
2022-01-11 12:40:33,038 INFO plugins.dependencies :: Plugin 'Metabase Presto JDBC Driver' depends on plugin 'Presto Common Driver'
2022-01-11 12:40:33,039 INFO plugins.dependencies :: Metabase Presto JDBC Driver dependency {:plugin Presto Common Driver} satisfied? true
2022-01-11 12:40:33,039 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :presto-jdbc...
2022-01-11 12:40:33,040 INFO driver.impl :: Registered driver :presto-jdbc (parents: [:presto-common]) 🚚
2022-01-11 12:40:33,043 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :google...
2022-01-11 12:40:33,044 INFO driver.impl :: Registered abstract driver :google  🚚
2022-01-11 12:40:33,044 INFO plugins.dependencies :: Metabase BigQuery Driver dependency {:plugin Metabase Google Drivers Shared Dependencies} satisfied? true
2022-01-11 12:40:33,045 INFO plugins.dependencies :: Metabase Google Analytics Driver dependency {:plugin Metabase Google Drivers Shared Dependencies} satisfied? true
2022-01-11 12:40:33,046 DEBUG plugins.initialize :: Dependencies satisfied; these plugins will now be loaded: ["Metabase Google Analytics Driver" "Metabase BigQuery Driver"]
2022-01-11 12:40:33,046 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :googleanalytics...
2022-01-11 12:40:33,047 INFO driver.impl :: Registered driver :googleanalytics (parents: [:google]) 🚚
2022-01-11 12:40:33,047 WARN plugins.lazy-loaded-driver :: Warning: plugin manifest for :bigquery does not include connection properties
2022-01-11 12:40:33,048 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :bigquery...
2022-01-11 12:40:33,049 INFO driver.impl :: Registered driver :bigquery (parents: [:sql :google]) 🚚
2022-01-11 12:40:33,071 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :hive-like...
2022-01-11 12:40:33,072 INFO driver.impl :: Registered abstract driver :hive-like (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:33,072 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :sparksql...
2022-01-11 12:40:33,073 INFO driver.impl :: Registered driver :sparksql (parents: [:hive-like]) 🚚
2022-01-11 12:40:33,077 INFO plugins.dependencies :: Metabase cannot initialize plugin Metabase Vertica Driver due to required dependencies. Metabase requires the Vertica JDBC driver in order to connect to Vertica databases, but we can't ship it as part of Metabase due to licensing restrictions. See https://metabase.com/docs/latest/administration-guide/databases/vertica.html for more details.

2022-01-11 12:40:33,078 INFO plugins.dependencies :: Metabase Vertica Driver dependency {:class com.vertica.jdbc.Driver} satisfied? false
2022-01-11 12:40:33,078 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase Oracle Driver" "Metabase Vertica Driver"]
2022-01-11 12:40:33,083 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :redshift...
2022-01-11 12:40:33,084 INFO driver.impl :: Registered driver :redshift (parents: [:postgres]) 🚚
2022-01-11 12:40:33,088 DEBUG plugins.lazy-loaded-driver :: Registering lazy loading driver :sqlite...
2022-01-11 12:40:33,089 INFO driver.impl :: Registered driver :sqlite (parents: [:sql-jdbc]) 🚚
2022-01-11 12:40:33,090 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
2022-01-11 12:40:33,095 INFO db.setup :: Verifying postgres Database Connection ...
2022-01-11 12:40:33,381 INFO db.setup :: Successfully verified PostgreSQL 13.4 (Ubuntu 13.4-4.pgdg18.04+1) application database connection. ✅
2022-01-11 12:40:33,382 INFO db.setup :: Running Database Migrations...
2022-01-11 12:40:33,404 DEBUG db.setup :: Set transaction to automatically roll back...
2022-01-11 12:40:33,404 DEBUG db.setup :: Disable auto-commit...
2022-01-11 12:40:33,404 INFO db.setup :: Setting up Liquibase...
2022-01-11 12:40:33,461 INFO db.setup :: Liquibase is ready.
2022-01-11 12:40:33,462 INFO db.liquibase :: Checking if Database has unrun migrations...
2022-01-11 12:40:34,833 INFO db.liquibase :: Database has unrun migrations. Waiting for migration lock to be cleared...
2022-01-11 12:40:34,845 INFO db.liquibase :: Migration lock is cleared. Running migrations...
2022-01-11 12:40:34,866 INFO db.setup :: Database Migrations Current ...  ✅
2022-01-11 12:40:34,886 INFO db.data-migrations :: Running all necessary data migrations, this may take a minute.
2022-01-11 12:40:34,925 INFO db.data-migrations :: Finished running data migrations.
Database setup took 1.8 s
2022-01-11 12:40:34,927 DEBUG db.connection-pool-setup :: Set default db connection with connection pool...
2022-01-11 12:40:35,119 DEBUG setting.cache :: Checking whether settings cache is out of date (requires DB call)...
2022-01-11 12:40:35,119 DEBUG setting.cache :: Refreshing Settings cache...
2022-01-11 12:40:35,133 INFO metabase.events :: Starting events listener: :metabase.events.dependencies/Dependencies 👂
2022-01-11 12:40:35,139 INFO metabase.events :: Starting events listener: :metabase.events.activity-feed/ActivityFeed 👂
2022-01-11 12:40:35,141 INFO metabase.events :: Starting events listener: :metabase.events.metabot-lifecycle/MetaBotLifecycle 👂
2022-01-11 12:40:35,142 INFO metabase.events :: Starting events listener: :metabase.events.last-login/LastLogin 👂
2022-01-11 12:40:35,142 INFO metabase.events :: Starting events listener: :metabase.events.notifications/Notifications 👂
2022-01-11 12:40:35,142 INFO metabase.events :: Starting events listener: :metabase.events.sync-database/Sync 👂
2022-01-11 12:40:35,143 INFO metabase.events :: Starting events listener: :metabase.events.driver-notifications/DriverNotifications 👂
2022-01-11 12:40:35,143 INFO metabase.events :: Starting events listener: :metabase.events.revision/Revisions 👂
2022-01-11 12:40:35,144 INFO metabase.events :: Starting events listener: :metabase.events.view-log/ViewLog 👂
2022-01-11 12:40:35,195 DEBUG metabase.task :: Loading tasks namespace: metabase.task.follow-up-emails
2022-01-11 12:40:35,217 DEBUG metabase.task :: Loading tasks namespace: metabase.task.send-anonymous-stats
2022-01-11 12:40:35,223 DEBUG metabase.task :: Loading tasks namespace: metabase.task.send-pulses
2022-01-11 12:40:35,241 DEBUG metabase.task :: Loading tasks namespace: metabase.task.sync-databases
2022-01-11 12:40:35,274 DEBUG metabase.task :: Loading tasks namespace: metabase.task.task-history-cleanup
2022-01-11 12:40:35,280 DEBUG metabase.task :: Loading tasks namespace: metabase.task.upgrade-checks
2022-01-11 12:40:35,327 INFO metabase.task :: Initializing task SyncDatabases 📆
2022-01-11 12:40:38,650 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:40:43,317 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:40:48,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:40:53,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:40:58,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:03,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:08,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:13,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:18,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:23,313 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:28,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:33,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:35,913 DEBUG setting.cache :: Checking whether settings cache is out of date (requires DB call)...
2022-01-11 12:41:38,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:43,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:48,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:53,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:41:58,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:03,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:08,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:13,315 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:18,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:23,313 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:28,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:33,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:36,910 DEBUG setting.cache :: Checking whether settings cache is out of date (requires DB call)...
2022-01-11 12:42:38,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:43,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:48,314 INFO middleware.exceptions :: Request canceled before finishing.
2022-01-11 12:42:53,314 INFO middleware.exceptions :: Request canceled before finishing.

...

2022-01-11 12:47:13,313 INFO middleware.exceptions :: Request canceled before finishing.
```

</details>

**Comment by bchrobot** [comment]
@flamber we're still struggling with this on v0.42.2. It seems like timing [this step](https://github.com/metabase/metabase/blob/085496d74602fb7a8e35d11e11f6f83abccf9382/src/metabase/task/sync_databases.clj#L281) and printing to debug would be a good next step. Any suggestions for best approach to do that?

**Comment by flamber** [comment]
@bchrobot All my databases are set to be [manual sync times](https://www.metabase.com/docs/latest/administration-guide/01-managing-databases.html#choose-when-metabase-syncs-and-scans), which is what you have to set, when I look at the [code](https://github.com/metabase/metabase/blob/085496d74602fb7a8e35d11e11f6f83abccf9382/src/metabase/task/sync_databases.clj#L255), so it doesn't try to randomize the schedule on startup.

Try starting with more logging - perhaps add other namespaces too:
```
-Dlog4j.configurationFile="https://log4j.us/v2/templates/metabase?trace=metabase.task"
```

I would probably also monitor the application database query log, since it can give you some insights into what/when Metabase is doing certain things.

**Comment by dpsutton** [comment]
Just a bit of clarification. The times are only randomized once if they look like the "old" default times. Previously, all of the databases would sync at one time, causing a heavy, punctuated load on the databases. It does not change the schedules on every startup once they are not the old defaults:

```clojure
schedules=> (map (juxt identity cron-util/describe-cron-string) default-metadata-sync-schedule-cron-strings)
(["0 50 * * * ? *" "At 50 minutes past the hour"]
 ["0 0 * * * ? *" "Every hour"])
schedules=> (map (juxt identity cron-util/describe-cron-string) default-cache-field-values-schedule-cron-strings)
(["0 0 0 * * ? *" "At 12:00 AM"] ["0 50 0 * * ? *" "At 12:50 AM"])
```

Once this is randomized once it won't randomize it again. And if the user has assumed control of scheduling it won't adjust them of course.

**Comment by bchrobot** [comment]
> All my databases are set to be manual sync times, which is what you have to set, when I look at the code, so it doesn't try to randomize the schedule on startup.

We tried setting all databases to manual sync using the statement below (unsure if this is correct) but still encountered the slow startup times.

```sql
update metabase_database set is_full_sync = false, is_on_demand = true;
```

> Once this is randomized once it won't randomize it again.

Subsequent startup times continued to be 5+ hours.

> Any updates on how to address this issue?

We ended up "solving" the problem by deleting a large number of databases and vacuuming the `qrtz_trigger` and `qrtz_cron_trigger` tables (we run Metabase on Postgres). Unfortunately we were not able to wait on trying each separately.

We noticed that even a count query on those tables was very slow. Vacuuming removed ~50 million dead tuples for the tables and each of the ~10 indexes. This seems consistent with queries to check schedules being responsible for the slow down as opposed to actually updating the schedule.

**Comment by flamber** [comment]
Try completely disabling sync (you can trigger manual sync via the API if needed) - https://github.com/metabase/metabase/issues/10398#issuecomment-514412107

**Comment by bchrobot** [comment]
We have since confirmed that Quartz indexes are the issue, specifically excessive indexes on `qrtz_triggers` (13 on the table!).

For such a frequently read/written table, this resulted in index fragmentation and blocked autovacuum processes. These in turn impacted performance of queries on the table (and thus the Metabase startup time) as well as sever storage bloat. Reindexing the table addressed performance and storage bloat issues.

One option is to reindex regularly:

```sql
select cron.schedule_in_database(
  '0 * * * *',
  'my_metabase_db',
  'reindex table concurrently qrtz_triggers'
);
```

Another is to drop the indexes altogether. The way we have observed Metabase using Quartz, there are rarely more than 1000 records in `qrtz_triggers`, so a sequential scan is just about as performant and heads off the index issues.

**Comment by paoliniluis** [comment]
@k-popov please open a new issue linked to this one

**Comment by dpsutton** [comment]
@k-popov I'd love to know the frequency of your databases. Something like

```sql
select engine, count(*) from metabase_database group by engine;
       engine       | count
--------------------+-------
 redshift           |     1
 snowflake          |     2
 postgres           |     4
 h2                 |     1
 bigquery-cloud-sdk |     2
 sqlite             |     3
 mysql              |     1
```

**Comment by k-popov** [comment]
> the frequency of your databases

@dpsutton , here you are:
```
metabase=> select engine, count(*) from metabase_database group by engine;
       engine       | count
--------------------+-------
 postgres           |   232
 bigquery-cloud-sdk |     1
 h2                 |     1
 mysql              |     6
 clickhouse         |    15
 sqlserver          |     2
(6 rows)
```
There is issue #52188 about clickhouse driver in metabase but that's not the case - the behaviour is the same without clickhouse driver.
